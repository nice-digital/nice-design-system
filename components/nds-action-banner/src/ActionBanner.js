import React from "react";
import PropTypes from "prop-types";

import RemoveIcon from "@nice-digital/icons/lib/Remove";

import "../scss/action-banner.scss";

export class ActionBanner extends React.Component {
	constructor(props) {
		super(props);

		this.state = {
			isClosed: false
		};

		this.closeClickHandler = this.closeClickHandler.bind(this);
	}

	closeClickHandler() {
		this.setState({
			isClosed: true
		});

		if (typeof this.props.onClosing === "function") this.props.onClosing(this);
		else throw new Error("The onClosing prop should be a function");
	}

	render() {
		if (this.state.isClosed) return null;

		const { variant, onClosing, title, children, cta } = this.props;

		const classes = ["action-banner"];

		if (variant && variant !== "default")
			classes.push(`action-banner--${variant}`);

		if (onClosing) classes.push("action-banner--closeable");

		return (
			<section className={classes.join(" ")}>
				<div className="action-banner__container">
					<div className="action-banner__inner">
						<div className="action-banner__text">
							<h2 className="action-banner__title">{title}</h2>
							{children && <p className="action-banner__intro">{children}</p>}
						</div>
						{cta && <div className="action-banner__actions">{cta}</div>}
						{onClosing && (
							<button
								type="button"
								className="action-banner__close"
								onClick={this.closeClickHandler}
							>
								<RemoveIcon />
								<span className="visually-hidden">Close {title}</span>
							</button>
						)}
					</div>
				</div>
			</section>
		);
	}
}

ActionBanner.propTypes = {
	title: PropTypes.string.isRequired,
	variant: PropTypes.oneOf(["default", "subtle"]),
	children: PropTypes.oneOfType([
		PropTypes.arrayOf(PropTypes.node),
		PropTypes.node
	]).isRequired,
	cta: PropTypes.node,
	onClosing: PropTypes.func
};

ActionBanner.defaultProps = {
	variant: "default"
};
