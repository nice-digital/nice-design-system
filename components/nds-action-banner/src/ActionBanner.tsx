import React, { ReactNode } from "react";
import classnames from "classnames";

import "../scss/action-banner.scss";

export interface ActionBannerProps {
	title: string;
	variant?: "default" | "subtle" | "fullWidth" | "fullWidthSubtle";
	children: ReactNode[] | ReactNode;
	cta: ReactNode;
	className?: string;
	image?: string;
}

export const ActionBanner: React.FC<ActionBannerProps> = (
	props
): JSX.Element | null => {
	const { variant, title, children, cta, className, image, ...rest } = props;

	const kebabCaseVariantClassName = variant
		? `${variant.replace(/([a-z])([A-Z])/g, "$1-$2").toLowerCase()}`
		: "";

	const classes = {
		"action-banner": true,
		[`action-banner--${kebabCaseVariantClassName}`]: variant,
		[`${className}`]: className
	};

	return (
		<section
			className={classnames(classes)}
			data-component={`action-banner${variant ? `--${variant}` : ""}`}
			{...rest}
		>
			{image && variant === "fullWidth" ? (
				<div
					className="action-banner--full-width__image-container"
					style={{ backgroundImage: `url(${image})` }}
				></div>
			) : null}
			<div className="action-banner__container">
				<div className="action-banner__inner">
					<div className="action-banner__text">
						<h2 className="action-banner__title">{title}</h2>
						{children && <div className="action-banner__intro">{children}</div>}
					</div>
					{cta && <div className="action-banner__actions">{cta}</div>}
				</div>
			</div>
		</section>
	);
};
