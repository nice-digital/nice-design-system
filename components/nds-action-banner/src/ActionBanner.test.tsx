import { render, screen } from "@testing-library/react";
import { ActionBanner } from "./ActionBanner";

describe("ActionBanner", () => {
	it("should render child text", () => {
		const text = "Some text";
		const wrapper = render(<ActionBanner title="Title">{text}</ActionBanner>);
		expect(screen.getByText(text)).toBeInTheDocument();
	});
});
