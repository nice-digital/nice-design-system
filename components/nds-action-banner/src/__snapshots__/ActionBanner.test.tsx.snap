// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionBanner should match snapshot for closeable subtle action banner 1`] = `
<div>
  <section
    class="action-banner action-banner--closeable action-banner--subtle"
  >
    <div
      class="action-banner__container"
    >
      <div
        class="action-banner__inner"
      >
        <div
          class="action-banner__text"
        >
          <h2
            class="action-banner__title"
          >
            Some title
          </h2>
          <p
            class="action-banner__intro"
          >
            Some body
          </p>
        </div>
        <div
          class="action-banner__actions"
        >
          <a
            href="/test"
          >
            Some CTA
          </a>
        </div>
        <button
          class="action-banner__close"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="icon"
            fill="none"
            focusable="false"
            height="1em"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M480 392.508c0 10.052-3.52 18.598-10.559 25.637l-51.296 51.296C411.106 476.48 402.56 480 392.508 480c-10.053 0-18.598-3.52-25.638-10.559L256 358.571l-110.87 110.87C138.09 476.48 129.545 480 119.492 480c-10.052 0-18.598-3.52-25.637-10.559l-51.296-51.296C35.52 411.106 32 402.56 32 392.508c0-10.053 3.52-18.598 10.56-25.638L153.428 256 42.559 145.13C35.52 138.09 32 129.545 32 119.492c0-10.052 3.52-18.598 10.56-25.637l51.295-51.296C100.894 35.52 109.44 32 119.492 32c10.053 0 18.598 3.52 25.638 10.56L256 153.428l110.87-110.87C373.91 35.52 382.455 32 392.508 32c10.052 0 18.598 3.52 25.637 10.56l51.296 51.295C476.48 100.894 480 109.44 480 119.492c0 10.053-3.52 18.598-10.559 25.638L358.571 256l110.87 110.87C476.48 373.91 480 382.455 480 392.508Z"
              fill="currentColor"
            />
          </svg>
          <span
            class="visually-hidden"
          >
            Close 
            Some title
          </span>
        </button>
      </div>
    </div>
  </section>
</div>
`;

exports[`ActionBanner should match snapshot for default action banner 1`] = `
<div>
  <section
    class="action-banner"
  >
    <div
      class="action-banner__container"
    >
      <div
        class="action-banner__inner"
      >
        <div
          class="action-banner__text"
        >
          <h2
            class="action-banner__title"
          >
            Some title
          </h2>
          <p
            class="action-banner__intro"
          >
            Some body
          </p>
        </div>
        <div
          class="action-banner__actions"
        >
          <a
            href="/test"
          >
            Some CTA
          </a>
        </div>
      </div>
    </div>
  </section>
</div>
`;
